### PR TITLE
Decouple parseable-output from the driver and sink it into the frontend

### PR DESCRIPTION
**Pull request by DougGregor**
Factor out the Driver-level dependencies on the parseable output generator and sink it down into the Frontend library. This is a step toward teaching the frontend to also emit parseable output, which is useful in the integrated Swift driver.

### CODE REVIEW

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by davidungar** [comment]
It's late & I'm tired. Is there an overall "why" comment somewhere, that explains that the purpose of having the frontend emit this format is so that Xcode (or xcbuild) can process the output for presentation in the IDE?

File: lib/Frontend/ParseableOutput.cpp [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+#include "swift/Basic/JSONSerialization.h"
+#include "swift/Basic/TaskQueue.h"
+#include "llvm/Option/Arg.h"
+#include "llvm/Support/raw_ostream.h"
+
+using namespace swift::parseable_output;
+using namespace swift::sys;
+using namespace swift;
+
+namespace swift {
```
Comment (davidungar) [inline lib/Frontend/ParseableOutput.cpp:25]: Some sort of overall comment would be nice.
Are all these traits teaching a JSON formatter how to output this stuff? (That's my sleepy guess.)

**Review by davidungar — APPROVED** [review]
I didn't think it was necessary to review this fully in depth. Just gave it a quick once-over. I made a few suggestions, but if you want to ship it as-is, it's fine.

**Comment by DougGregor** [comment]
> It's late & I'm tired, so maybe it's there already. Is there an overall "why" comment somewhere, that explains that the purpose of having the frontend emit this format is so that Xcode (or xcbuild) can process the output for presentation in the IDE?

I'll add a short comment to the header documentation.

File: lib/Driver/Job.cpp [hunk 2]
```diff
@@ -450,6 +452,54 @@ StringRef Job::getFirstSwiftPrimaryInput() const {
   return StringRef();
 }
 
+Job::operator JobInfo() const {
```
[the remark this answers is not in the document]
Comment (DougGregor) [reply lib/Driver/Job.cpp:455]: It's a user-defined conversion. It needs to be written this way for the layering to work out: the `Job` is in the Driver, `JobInfo` is in the Frontend, and Driver links to Frontend.

File: lib/Frontend/ParseableOutput.cpp (hunk 1 above)
Comment (DougGregor) [reply lib/Frontend/ParseableOutput.cpp:25]: Yes, this is why these traits are defined in the `json` namespace

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci smoke test

**Comment by DougGregor** [comment]
@swift-ci please smoke test

File: lib/Driver/Job.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (davidungar) [reply lib/Driver/Job.cpp:455]: Ah! Thanks. The 2nd sentence would make a great comment for newbies in code base.

File: lib/Frontend/ParseableOutput.cpp (hunk 1 above)
Comment (davidungar) [reply lib/Frontend/ParseableOutput.cpp:25]: If you needed to extend the parsable output in a way that involved these traits, and you were brand-new to the code, would a comment here help? Or do you think it's just obvious?

**Comment by DougGregor** [comment]
@swift-ci smoke test

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci clean smoke test Linux

**Comment by shahmishal** [comment]
Please update the base branch to `main` by Oct 5th otherwise the pull request will be closed automatically. 

* How to change the base branch: ([Link](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-base-branch-of-a-pull-request))
* More detail about the branch update: ([Link](https://forums.swift.org/t/updating-branch-names/40412/))

**Pull request closed without merging**